crypto/src/merkle_tree/mod.rs

```diff
@@ -1,187 +1,258 @@
+        .map(|val| hasher.hash_one(val.clone()))
         .collect()
 }
 
-fn hash_level<F: IsField, H: IsCryptoHash<F>>(
-    values: &mut Vec<TreeNode<F>>,
-    hasher: &H,
-) -> Vec<TreeNode<F>> {
-    if values.len() % 2 != 0 {
-        values.push(values[values.len() - 1].clone());
+// The list of values is completed repeating the last value to a power of two length
+fn complete_until_power_of_two<F: IsField>(
+    values: &mut Vec<FieldElement<F>>,
+) -> Vec<FieldElement<F>> {
+    while !is_power_of_two(values.len()) {
+        values.push(values[values.len() - 1].clone())
     }
+    values.to_vec()
+}
 
-    values
-        .chunks(2)
-        .map(|chunk| build_parent(chunk[0].clone(), chunk[1].clone(), hasher))
-        .collect()
+fn is_power_of_two(x: usize) -> bool {
+    (x != 0) && ((x & (x - 1)) == 0)
 }
 
-fn build_parent<F: IsField, H: IsCryptoHash<F>>(
-    left: TreeNode<F>,
-    right: TreeNode<F>,
+fn build<F: IsField, H: IsCryptoHash<F>>(
+    nodes: &mut Vec<FieldElement<F>>,
+    parent_index: usize,
     hasher: &H,
-) -> TreeNode<F> {
-    let parent_hash = hasher.hash_two(left.borrow().hash.clone(), right.borrow().hash.clone());
-    let parent = build_tree_node(parent_hash);
+) -> Vec<FieldElement<F>> {
+    if is_leaf(nodes.len(), parent_index) {
+        return nodes.to_vec();
+    }
 
-    left.borrow_mut().sibiling = Some(right.clone());
-    left.borrow_mut().parent = Some(parent.clone());
+    let left_child_index = left_child_index(parent_index);
+    let right_child_index = right_child_index(parent_index);
 
-    right.borrow_mut().sibiling = Some(left);
-    right.borrow_mut().parent = Some(parent.clone());
+    let mut nodes = build(nodes, left_child_index, hasher);
+    nodes = build(&mut nodes, right_child_index, hasher);
 
-    parent
+    nodes[parent_index] = hasher.hash_two(
+        nodes[left_child_index].clone(),
+        nodes[right_child_index].clone(),
+    );
+    nodes
 }
 
-fn build_merkle_path<F: IsField>(
-    node: TreeNode<F>,
-    merkle_path: &mut Vec<TreeNode<F>>,
-) -> Vec<TreeNode<F>> {
-    merkle_path.push(node.clone());
+fn is_leaf(lenght: usize, node_index: usize) -> bool {
+    (node_index >= (lenght / 2)) && node_index < lenght
+}
 
-    if let Some(parent) = node.borrow().parent.clone() {
-        return build_merkle_path(parent, merkle_path).to_vec();
-    }
+fn left_child_index(parent_index: usize) -> usize {
+    parent_index * 2 + 1
+}
 
-    merkle_path.to_vec()
+fn right_child_index(parent_index: usize) -> usize {
+    parent_index * 2 + 2
 }
 
 pub struct Proof<F: IsField, H: IsCryptoHash<F>> {
     value: FieldElement<F>,
-    merkle_path: Vec<TreeNode<F>>,
+    merkle_path: Vec<(FieldElement<F>, bool)>,
     hasher: H,
 }
 
-type TreeNode<F> = Rc<RefCell<Node<F>>>;
+pub type F = U64PrimeField<0xFFFF_FFFF_0000_0001_u64>;
+pub type FE = FieldElement<F>;
+
+pub type U64MerkleTree = MerkleTree<F, DefaultHasher>;
+
+impl Proof<F, DefaultHasher> {
```

Comment:
We should implement `ByteConversion` instead of implementing `as_bytes` and `from_bytes` alone.